storage/innobase/xtrabackup/src/xtrabackup.cc

```
@@ -3366,16 +3374,25 @@ static void data_copy_thread_func(data_thread_ctxt_t *ctxt) {
 
   while ((node = datafiles_iter_next(ctxt->it)) != NULL && !*(ctxt->error)) {
     /* copy the datafile */
-    if (xtrabackup_copy_datafile(node, num)) {
+    bool is_error = xtrabackup_copy_datafile(node, num);
+
+    if (!is_error) {
+      if (ddl_tracker != nullptr && opt_lock_ddl == LOCK_DDL_REDUCED) {
+        /* With reduced lock mode, instead of tracking undo from the startup
+          scan, we track undo tablespaces after we copy them */
+        if (fsp_is_undo_tablespace(node->space->id)) {
+          ddl_tracker->add_undo_tablespace(node->space->id,
+                                           node->space->files.front().name);
+        } else if (fsp_is_ibd_tablespace(node->space->id)) {
+          ddl_tracker->add_table_from_ibd_scan(node->space->id,
+                                               node->space->files.front().name,
+                                               node->space->flags);
+        }
```

I would an else branch. this will be unhandled path? On release build, print message about space->id, name and return error.

On debug build, let it crash ut_ad(0);

---

Do you mean we need to crash in `else case` (not fsp_is_undo_tablespace and not fsp_is_ibd_tablespace) ?

---

yes, we should hanlde all cases So am talking about the else branch
if () {
} else if () {
} else {
--> unexpected path, so crash here on debug builds and  print warning for releae builds
}

---

left this as is, in else case we have all other tablespaces like system/tmp etc